chain/epoch-manager/src/validator_selection.rs

```diff
@@ -33,49 +123,37 @@ pub fn proposals_to_epoch_info(
     );
 
     let shard_ids: Vec<_> = epoch_config.shard_layout.shard_ids().collect();
-    let min_stake_ratio = {
-        let rational = epoch_config.validator_selection_config.minimum_stake_ratio;
-        Ratio::new(*rational.numer() as u128, *rational.denom() as u128)
-    };
-    let max_bp_selected = epoch_config.num_block_producer_seats as usize;
     let mut stake_change = BTreeMap::new();
-    let proposals = proposals_with_rollover(
+    let proposals = apply_epoch_update_to_proposals(
         proposals,
         prev_epoch_info,
         &validator_reward,
         &validator_kickout,
         &mut stake_change,
     );
-    let mut block_producer_proposals = order_proposals(proposals.values().cloned());
-    let (block_producers, bp_stake_threshold) = select_block_producers(
-        &mut block_producer_proposals,
-        max_bp_selected,
-        min_stake_ratio,
-        current_version,
-    );
-    let (chunk_producer_proposals, chunk_producers, cp_stake_threshold) =
-        if checked_feature!("stable", ChunkOnlyProducers, next_version) {
-            let mut chunk_producer_proposals = order_proposals(proposals.into_values());
-            let max_cp_selected = max_bp_selected
-                + (epoch_config.validator_selection_config.num_chunk_only_producer_seats as usize);
-            let (chunk_producers, cp_stake_threshold) = select_chunk_producers(
-                &mut chunk_producer_proposals,
-                max_cp_selected,
-                min_stake_ratio,
-                shard_ids.len() as NumShards,
-                current_version,
-            );
-            (chunk_producer_proposals, chunk_producers, cp_stake_threshold)
-        } else {
-            (block_producer_proposals, block_producers.clone(), bp_stake_threshold)
-        };
 
-    // since block producer proposals could become chunk producers, their actual stake threshold
-    // is the smaller of the two thresholds
-    let threshold = cmp::min(bp_stake_threshold, cp_stake_threshold);
+    // Select validators for the next epoch.
+    // Returns unselected proposals, validator lists for all roles and stake
+    // threshold to become a validator.
+    let ValidatorRoles {
+        unselected_proposals,
+        chunk_producers,
+        block_producers,
+        chunk_validators,
+        threshold,
+    } = if checked_feature!("stable", StatelessValidationV0, next_version) {
+        select_validators_from_proposals(epoch_config, proposals, next_version)
+    } else {
+        old_validator_selection::select_validators_from_proposals(
```

Comment:
very nit. maybe not say "old" yet but something like stateful_validator_selection, as this will still be effective until the launch? or V1 vs V2 as in the case of other evolving structs in the code?

Reply:
I would keep `old_validator_selection` for now because this is existing naming for old versions and easy to fix in the future.